lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -284,538 +352,512 @@ bool PatchEntryPointMutate::isResourceNodeActive(const ResourceNode *node, bool
+//
+// * (For a GFX9+ merged shader or NGG primitive shader, the 8 system SGPRs at the start are not accounted for here.)
+// * The "user data" SGPRs, up to 32 (GFX9+ non-compute shader) or 16 (compute shader or <=GFX8). Many of the values
+//   here are pointers, but are passed as a single 32-bit register and then expanded to 64-bit in the shader code:
+//   - The "global information table", containing various descriptors such as the inter-shader rings
+//   - The "per-shader table", which is added here but appears to be unused
+//   - The streamout table if needed
+//   - Nodes from the root user data layout, including pointers to descriptor sets. In a compute shader, these
+//     must have the same layout as the root user data layout, even if that would mean leaving gaps in register
+//     usage, and are limited to 10 SGPRs s2-s11
+//   - For a compute shader, the spill table pointer if needed, which always goes into s12. (This is possibly
+//     not a strict requirement of PAL, but it might avoid extraneous HW register writes.)
+//   - For a compute shader, the NumWorkgroupsPtr register pair if needed, which must start at a 2-aligned
+//     register number, although it can be earlier than s14 if there is no spill table pointer
+//   - Various other system values set up by PAL, such as the vertex buffer table and the vertex base index
+//   - For a graphics shader, the spill table pointer if needed. This is typically in the last register
+//     (s15 or s31), but not necessarily.
+// * The system value SGPRs and VGPRs determined by hardware, some of which are enabled or disabled by bits in SPI
+//   registers.
+//
+// @param [out] inRegMask : "Inreg" bit mask for the arguments, with a bit set to indicate that the corresponding
+//                          arg needs to have an "inreg" attribute to put the arg into SGPRs rather than VGPRs
+// @return : The newly-constructed function type
 //
-// @param [out] inRegMask : "Inreg" bit mask for the arguments
 FunctionType *PatchEntryPointMutate::generateEntryPointType(uint64_t *inRegMask) const {
-  unsigned userDataIdx = 0;
-  std::vector<Type *> argTys;
+  SmallVector<Type *, 8> argTys;
 
-  auto userDataNodes = m_pipelineState->getUserDataNodes();
+  IRBuilder<> builder(*m_context);
   auto intfData = m_pipelineState->getShaderInterfaceData(m_shaderStage);
-  auto resUsage = m_pipelineState->getShaderResourceUsage(m_shaderStage);
+  auto &entryArgIdxs = intfData->entryArgIdxs;
+  entryArgIdxs.initialized = true;
+
+  // First we collect the user data args in two vectors:
+  // - userDataArgs: global table, per-shader table and streamout table, followed by the nodes from the root user
+  //   data layout (excluding vertex buffer and streamout tables). Some of these are marked mustSpill; some of them
+  //   may need to be spilled anyway due to running out of entry SGPRs
+  // - extraUserDataArgs: special values that go at the end, such as ViewId.
```

Comment:
why do we need put ViewId in the end of user datas? is it limited by PAL or just to simplify the design?
If I remember correctly, one reason is some special user data can't be put in spill table in PAL. so what will happen if we put these in the beginning of user data?

Reply:
I don't know; I just copied what it did before. :-) I suspect it was to ensure it is in the same register in two shaders that are merged.

Putting at the beginning of user data should be fine in terms of not spilling. The global table and streamout table already do that.

But I propose not changing anything like that in this change.